ruby/sdk-phase-2.md

```diff
@@ -0,0 +1,575 @@
+
+    # Called in the workflow definition to change the name from the default.
+    def self.workflow_name(name)
+
+    # Placed above initialize to get start arguments.
+    def self.workflow_init
+
+    # See dynamic section.
+    def self.workflow_dynamic
+
+    # Called in the workflow definition for initialize and execute to get
+    # RawValue arguments (useful for dynamic workflows).
+    def self.workflow_raw_args
+
+    ########
+    # Class methods that can only be called during runtime, but don't have to be
+    # in a workflow context
+    ########
+
+    # Whether in a workflow (can only be called on Workflow class, not child).
+    def self.in_workflow?
+
+    # Built, fixed workflow definition (can only be called on a class extending
+    # workflow, not workflow itself).
+    def self.definition
+
+    ########
+    # Class methods that can only be called during runtime in a workflow context
+    ########
+
+    # No start_activity because that can be in a future (see asynchrony section).
+    def self.execute_activity(activity, *args, **left_off_for_brevity)
+    def self.execute_local_activity(activity, *args, **left_off_for_brevity)
+
+    # Returns ChildWorkflowHandle.
+    def self.start_child_workflow(workflow, *args, **left_off_for_brevity)
+    # Returns actual result.
+    def self.execute_child_workflow(workflow, *args, **left_off_for_brevity)
+
+    # These are a custom hash class that only supports []= mutation and
+    # validates input as a signal/query/update definition. Basically they will
+    # be simple delegators to frozen hashes only updatable via index. Symbol
+    # keys or nil key for dynamic. All values are "definition" objects.
+    def self.signal_handlers
+    def self.query_handlers
+    def self.update_handlers
+
+    def info
+    def current_update_info
+
+    def self.search_attributes
+    def self.upsert_search_attributes(*updates)
+    def self.raw_memo
+    def self.memo_value(name)
+    def self.upsert_memo(*updates)
+
+    def self.external_workflow_handle(id, run_id: nil)
+
+    def self.patched(id)
+    def self.deprecate_patch(id)
+
+    def self.cancellation
+    # Just a shortcut to getting a future that fails on cancel. Good for
+    # Future.any_of use.
+    def self.cancellation_future
+    def self.logger
+    def self.now
+    def self.payload_converter
+    def self.random
+    def self.uuid
+    def self.wait_condition(&)
+
+    class ContinueAsNewError < Error
+      def initialize(*args, workflow: nil, **left_off_for_brevity)
+    end
+
+    class Unsafe
+      def replaying?
+      # See TracePoint later on why this is present.
+      def illegal_call_tracing_disabled(&)
+    end
+
+    # Other nested classes omitted for brevity
+  end
+end
+```
+
+Notes:
+
+* Notice lack of `timeout` on `wait_condition`, we will support Ruby `Timeout` via the fiber scheduler.
+* Notice lack of `sleep`, we will support Ruby `sleep` via the fiber scheduler.
```

Comment:
Hrmm. Ruby explicitly delegates these two things to the scheduler: https://docs.ruby-lang.org/en/master/Fiber/Scheduler.html#method-i-timeout_after and https://docs.ruby-lang.org/en/master/Fiber/Scheduler.html#method-i-kernel_sleep.

The concern with erroring on these is that we are probably going to recommend use of the https://github.com/socketry/async library (and definitely heavily test for its use) because Ruby doesn't have good async primitives and that library leans on existing fibers. So users expect to be able to call https://socketry.github.io/async/source/Async/Task/index.html#Async::Task#with_timeout and https://socketry.github.io/async/source/Async/Task/index.html#Async::Task#sleep (now deprecated because they too encourage stdlib sleep).

This is the same type of thing with `Queue` where Ruby built in support for customized async inside it so people can use it without concern. I don't expect us to say "queue.pop confuses users and we should have our own queue" either.

In this case, I don't think these are considered "magic patched", I think it's expected by users that they work.

> If the main counter argument is that you can re-use some existing code, I know I've heard agreement from you and others before that that's very rare to actually be viable.

I think that's true elsewhere, but in Ruby, the lack of async primitives means we have to either write a ton of them, or allow use of the common `async` library built on the same fiber abstraction we're building on (almost all of that library's primitives will work based on research so far). But open to options here. I think we need to pick between:

* Allow/encourage use of https://github.com/socketry/async
* Write our own primitives (e.g. Mutex and notification and such) and disallow/discourage use of that library

We already have the basic future primitive, so technically the second option is viable, but the proposal currently is expecting the first. (pardon the long response)

Reply:
OK, yeah I can buy the argument about socketry using these under the hood.

Reply:
See https://github.com/temporalio/proposals/pull/96#discussion_r1817109910, maybe we don't want to encourage this `async` lib after all. It does put too much trust in their deterministic fiber scheduler use (even across versions).

Reply:
Clarified that we don't want to encourage `async` but the stdlib primitives that work in all concurrent environments (i.e. thread and fiber) can be used, but we will alias/wrap them.